src/view/reviewManager.ts

```diff
@@ -76,8 +82,30 @@ export class ReviewManager implements vscode.DecorationProvider {
 			vscode.commands.executeCommand('vscode.open', vscode.Uri.file(nodePath.resolve(this._repository.path, params.path)), opts);
 		}));
 		this._disposables.push(_repository.onDidRunGitStatus(e => {
-			// todo, validate state only when state changes.
-			this.updateState();
+			const oldHead = this._previousRepositoryState.HEAD;
+			const newHead = this._repository.HEAD;
+			const sameUpstream = !!oldHead.upstream
+				? newHead.upstream && oldHead.upstream.name === newHead.upstream.name && oldHead.upstream.remote === newHead.upstream.remote
+				: !!newHead.upstream;
+			const sameHead = oldHead.ahead === newHead.ahead
+				&& oldHead.behind === newHead.behind
+				&& oldHead.commit === newHead.commit
+				&& oldHead.name === newHead.name
+				&& oldHead.remote === newHead.remote
+				&& oldHead.type === newHead.type
+				&& sameUpstream;
+
+			const sameRemotes = this._previousRepositoryState.remotes.length === this._repository.remotes.length
+				&& this._previousRepositoryState.remotes.every(remote => this._repository.remotes.some(r => remote.equals(r)));
+
+			if (!sameHead || !sameRemotes) {
+				this._previousRepositoryState = {
+					HEAD: Object.assign(this._repository.HEAD),
```

Comment:
Is object duplication necessary here? `this._previousRepositoryState` will keep reference to the old `HEAD` as when `this._repository` status gets updated, we will create new objects for `HEAD` and `remotes`. Correct me if I miss any potential data leak.

Reply:
Nice, you're right, no need to duplicate